api/src/components/bookmark/controller.ts

```diff
@@ -1,114 +1,214 @@
-                message: 'This publication does not exist.'
-            });
-        }
+        // Check that the bookmark exists
+        const bookmark = await bookmarkService.get(event.pathParameters.id);
 
-        //check that the publication is live
-        if (publication.currentStatus === 'DRAFT') {
+        if (!bookmark) {
             return response.json(404, {
-                message: 'You cannot bookmark a draft publication.'
+                message: 'This bookmark does not exist.'
             });
         }
 
-        const bookmark = await bookmarkService.get(event.pathParameters.id, event.user.id);
+        const { id, type, entityId } = bookmark;
 
-        // check that the user has created the bookmark/ hasn't already deleted it
-        if (!bookmark) {
-            return response.json(404, {
-                message: 'You have already deleted the bookmark for this publication.'
+        // Check that the user owns the bookmark
+        if (bookmark.userId !== event.user.id) {
+            return response.json(403, {
+                message: 'You do not own this bookmark.'
             });
         }
 
-        await bookmarkService.remove(bookmark.id);
+        // Checks depending on bookmark type
+        switch (type) {
+            case 'PUBLICATION': {
+                // Check that the publication exists
+                const publication = await publicationService.get(entityId);
+
+                if (!publication) {
+                    return response.json(404, {
+                        message: 'This publication does not exist.'
+                    });
+                }
+
+                // Check that the publication is live
+                if (publication.currentStatus === 'DRAFT') {
+                    return response.json(404, {
+                        message: 'You cannot bookmark a draft publication.'
+                    });
+                }
+
+                break;
+            }
+
+            case 'TOPIC': {
+                // Check that the topic exists
+                const topic = await topicService.get(entityId);
+
+                if (!topic) {
+                    return response.json(404, {
+                        message: 'This topic does not exist.'
+                    });
+                }
+
+                break;
+            }
+
+            default:
+                // Invalid bookmark type supplied
+                return response.json(400, {
+                    message: 'Invalid bookmark type.'
+                });
+        }
+
+        // Delete bookmark
+        await bookmarkService.deleteBookmark(id);
 
-        return response.json(200, { message: 'Bookmark deleted from this publication.' });
+        return response.json(200, { message: 'Bookmark deleted successfully.' });
     } catch (err) {
         return response.json(500, { message: 'Unknown server error.' });
     }
 };
 
+// For checking whether an entity is bookmarked.
 export const get = async (
```

Comment:
this controller should simply return a bookmark by a given id. We can use getAll with some filters in order to check if a specific bookmark exists

Reply:
Removed this function as the UI doesn't need it anymore.